### DISCUSSION TITLE
RDF4j V5 should we require Java Release 11, 17 or something else?

### DISCUSSION THREAD

**Discussion by JervenBolleman** [eclipse-rdf4j/rdf4j#4380] [Polls]
For RDF4j version 5 we have the option to move our baseline JDK to 17.

This is annoying for some of us, but allows quite a few new features and some parts of the ecosystem are leading the way (e.g. spring-boot). We should also consider that premier support for java 11 ends sometime in the next 8 months (September 2023).

**Comment by barthanssens** [comment]
Using 21 is too soon IMHO, most organizations I talked to are just (thinking about) moving towards JDK 17.

**Reply by JervenBolleman** [reply]
Just as a note: JDK 21 won't be out until September 2023.

**Comment by hmottestad** [comment]
I thought we talked this over in the online meeting we had in the autumn and decided to stick with 11 for the time being.

Even though spring boot has moved to 17 for their 3.0.0 release they are still supporting 2.7.x until November this year for free and until 2025 for paying users.

**Reply by JervenBolleman** [reply]
Yes, but my recollection was different, that for v5 we would go to 17. Main reason for being that we want to release in early 2023, but by late 2023 v11 is out of active support for many users that don't have a commercial support contract. Which does not mean that they won't get updates, but that on new systems they might not have an jdk 11 to start with. This is not that important considering we do and will work with 17.

**Reply by erikgb** [reply]
Even if Spring Boot 2.x is still supported, we won't be able to migrate fully from javax -> jakarta without upgrading to Spring Boot 3.x. It may not be possible to have a mix of javax and jakarta.

**Reply by hmottestad** [reply]
Isn't the spring-boot stuff we have fairly disconnected from the rest of the code?

My point with spring-boot 2.7.x being supported is that they have the resources and money to continue supporting both Java 11 and 17 at the same time, while we have to choose.

**Comment by abrokenjester** [comment]
Is there anything we want to do that _requires_ Java 17? I know @erikgb has been looking at upgrading to spring-boot 3.x, but I'm not sure if that is a "would be nice" or a "can't use without it".

In any case I agree that as much as I'd personally like to get Java 17 features in our code, we have to be a little conservative to keep supporting existing projects.

If we did switch to Java 17 for 5.0, what features and improvements would projects stuck on Java 11 miss out on? Could we possibly support most of them by doing future 4.x releases in parallel for a while? Not for full feature parity perhaps, but for bug fixes and _some_ backported improvements at least.

**Reply by erikgb** [reply]
I would consider this a "can't use without it" as we see a bunch of project dependencies doing the javax -> jakarta migration, and that migrating is a DO or DON'T. No option for gradual migration AFAIK.

**Reply by abrokenjester** [reply]
@erikgb can you clarify why moving to Jakarta mandates this? AFAIK Jakarta 10 is still compatible with Java 11...  Admittedly I'm _wildly_ ignorant of the details of Jakarta EE and how it all fits together, though.

**Reply by erikgb** [reply]
@jeenbroekstra most of our application modules are based on spring/spring-boot, and this blog post summarizes their plans and actions: https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6. But when I think more about this, sorry for the confusion, I realize that the javax -> jakarta migration is really what we are looking for. Maybe Java 17 is a more "would be nice" requirement after all. 🤔

**Reply by abrokenjester** [reply]
Fwiw things are moving a lot faster in this space than they used to do. I would bet that if even we stick to Java 11 for now, RDF4J 6.0 will move to Java 17, and depending on urgency is something that we can decide to put up for release quite quickly.

**Reply by JervenBolleman** [reply]
I would really like to have records for use in the code base.

**Comment by hmottestad** [comment]
If Loom makes it into Java 21 I bet that a lot of web related frameworks and servers like tomcat etc. are going to move to that very swiftly. 

Loom has been hyped for a number of years now. It will bring very lightweight threads to Java. Currently a preview feature in Java 19. 

We might end up moving to Java 21 and skipping Java 17 if we don't upgrade this time around.
